### DISCUSSION TITLE
Learner vs Executor which one to customize?

### DISCUSSION THREAD

**Discussion by Jeffwan** [NVIDIA/NVFlare#1177] [Q&A] [answered]
I am learning NVFlare examples and try to custom my own executors. I meet a problem on the high level concepts

1. In hello_pt example, `Cifar10Trainer` is a subclass of Executor. https://nvflare.readthedocs.io/en/main/examples/hello_pt.html
2. In hello_pt_tb example, a `LearnerExecutor` and a custom `Learner` -> `PTLearner` are used.  https://nvflare.readthedocs.io/en/main/examples/hello_pt_tb.html

1. What's the difference? Should I custom Executor directly or custom `LearnerExecutor`? 
2. Seems Learner Executor covers `Validation` phase and I feel it's a high level abstraction of machine learning flow? If previous example, user need to define a separate `Cifar10validator.py` which is already covered by `Learner` api?

![image](https://user-images.githubusercontent.com/4739316/205722096-ba978cf3-146f-4ca3-93d1-206f8a763f23.png)

**Comment by chesterxgchen** [comment, marked as answer]
The Executor is the base (lower level) class that client need to implements. With this class, user need to handle raw task input from the server counter part (controller) and response with the results to the Controller. 

If you are design a branch new Algorithms, you might start with Controller on the server side and Executor on the client side.  Considering design Request (Controller) and Response (Client) 

This means, you have to deal with many things low level such as handling the communication, receiving results, handling errors and many NVFLARE specific events etc.  

For most of Machine learning (especially Deep Learning) pattern, these executor execution is the same, the only difference is the training, data, validation, input/out. In this case, you should use Learner Interface. This abstract away the communication with Flare Controller part and let you focused on the thing you care most, training/validation etc. Essentially hide away the complexity from end user. 

Before Learner, you would do the followings
```
      Server                Client side / User facing
    <Controller>  <-->     <Executor> 
```
With Learner 
```
      Server                 Client side           User facing 
    <Controller>  <-->    <Executor>  --> <Learner>
```
Hope this make explains.
